SDWebImage/SDWebImageDownloaderOperation.m

```diff
@@ -360,8 +361,10 @@ - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)data
         
         // progressive decode the image in coder queue
         dispatch_async(self.coderQueue, ^{
-            // If all the data has already been downloaded, earily return to avoid further decoding
-            if (self.receivedSize >= self.expectedSize) {
+            // Get the finish status when progressive decoding task actually executes
+            BOOL finished = (self.receivedSize >= self.expectedSize);
+            // If at the beginning of decoding task was dispatched, the download was not finished. But by the time it executes, the download is finished, then earily return to avoid further decoding.
+            if (downloadFinished != finished) {
```

Comment:
@dreampiggy 👌 I get your point, and you are more familiar than me in image format, as a layman, I just think if we already get all the data, why we still should do progressive data, I think we should only decode it in `- URLSession:task:didCompleteWithError:`. Let's wait other guys to review, I need to learn some image format firstly. 😂

Reply:
Actually...I think progressive decoding code logic here contains aother issue as well.

For example, since we separate the decoding task to coder queue, network callback to URLSession queue. I think that `[imageData copy]` should also be considered about queue change.

Current implementation, since we copy the image data in the URLSession delegate queue, so we will decode the data snapshot each time URLSession delegate callback.

However, what about, copy the imageData, just when the actual decoding task begin ? Which means we will decode the data snapshot each time coder queue run. This two are different and the second one will more accurate and less frequency, which represent current downloaded image data status.

Note this must use lock to ensure thread-safe.

```objective-c
dispatch_async(self.coderQueue, ^{
    // LOCK
    NSData *imageData = [self.imageData copy];
    // UNLOCK
    // decode current image data
    [self.progressiveDecoder decodeWithData:imageData];
};
```

Reply:
@dreampiggy Good catch, seems we have two choices:
1. Like you said, use `LOCK` to prevent race condition. Besides `imageData`, I think we also need let `self.receivedSize` safe, because we set in `URLSession` delegate but read in `coderQueue`.
2. Other is make all related variables to point to local variables, that is, put `imageData` copy in `NSURLSession` delegate method, and change `self.receivedSize` in `coderQueue` point to local variable, e.x. 

```
        // Get the image data
        NSData *imageData = [self.imageData copy];
        NSUInteger receivedSize = self.receivedSize;
        
        // progressive decode the image in coder queue
        dispatch_async(self.coderQueue, ^{
            // If all the data has already been downloaded, earily return to avoid further decoding
            if (receivedSize >= self.expectedSize) {
                return;
            }
```

Reply:
@zhongwuzw Correct. Both of two policy are adoptable, I don't know which one is better. So I say it's need discussion. 😕 

But since this discussion need other consideration, I can revert this changes just back to 4.x current code. Only fix the bug firstly...

Reply:
@dreampiggy Agree.

Reply:
@zhongwuzw Rebase to one commit.

This change roll back to 4.x behavior. We copy the image data snapshot each URLSession delegate call, as well as current finish status.

I think we can create a issue firstly to talk about whether we need to choose some other way for this.